### ISSUE TITLE
Traditional ſpacing as a ſtyleſet and or axis.

### ISSUE THREAD

**Issue by Alice196498** [psb1558/Elstob-font#26]
In traditional print, an em is uſed after terminal punctuation ; that is, the period and marks of interrogation and exclamation when ending a ſentence. Given that there is little way to determine a ſentence programmatically, without affecting abbreviations, which ſhould have no greater ſpacing after them, this may be difficult to implement ſufficiently error-free.
 A thin ſpace is always uſed before a ſemi-colon, colon, mark of interrogation, mark of exclamation, after opening quotation marks both ſingle and double, and before cloſing quotation marks both ſingle and double when other punctuation is not preſent before this cloſing mark. When a double and ſingle quotation mark fall directly beſide each other, they are likewiſe ſeparated by a thin ſpace. One difficulty here, and the reaſon for which we cannot ſimply, as a rule, have alternative marks with greater ſpace, is that the ſingle cloſing quotation mark is the very ſame character as the apoſtrophe, and the apoſtrophe is never ſeparated from its word by any ſpace.
 An en quad is ſet after a ſemi-colon, colon, and non-terminal marks of interrogation and exclamation, except in the caſe where a capital letter follows the colon—here, two thick ſpaces are uſed.

In general, theſe are the ſpaces that are taught to be uſed, but in practice, they more repreſent a ratio of ſpace ſurrounding punctuation, as they may be ſomewhat expanded or looſened depending on the needs of juſtification, but the ratio ſhould be generally perſerved. While it may ſomewhat go againſt the traditional rules of fixed ſpaces around punctuation, one may, perhaps, generalize the rules thus : Semi-colons, colons, marks of interrogation and exclamation, ſingle and double quotation marks require a ſmaller ſpace than the word ſpace to ſeparate them ſufficiently from the matter, and the former four require a wider ſpace than the word ſpace after them. Terminal punctuation requires an even wider ſpace ſtill.
 One thing that was done on the LinoType machine was to ſet an en quad and a ſpaceband after terminal punctuation, and a thin ſpace and ſpaceband after colons and ſemicolons, which allowed a relatively ſimilar ratio of a wider ſpace where needed, but as ſuch that it could alſo expand greater as needed. A thin ſpace was ſtill always uſed in the above caſes.
 Some printers alſo recommend a hair-ſpace ſurrounding em daſhes. Thin ſpaces are ſet around en daſhes that come between numerals.

With a ſtyleſet, theſe rules could be ſet into place, with certain exceptions built in to try and catch things like common abbreviations or caſes of apoſtrophes that are eaſily confuſed for ſingle quotation marks—it would not be perfect, but it ſhould be poſſible to have it work in a great majority of, if not nearly all, caſes.
 An axis may have difficulty accounting for exceptions, which would make it non-ideal, but does have the ability to change the width ſomewhat in accordance with the need of the compoſitor. I have alſo ſeen ſome articles which ſpeak of a [juſtification table](https://typedrawers.com/discussion/3465/making-jstf-better), which could be of potential uſe as well.

**Comment by psb1558** [comment]
This may be too complex a problem to handle with OT code, but let me lay out, as a starting position, what I did with Junicode. `ss18` substitutes alternate forms of **! ( ) : ; ? ‘ ’ “ ”** with extra space on the left side of punctuation marks and on the inside of bracketing characters.

Later on, `calt` tries to detect instances of quoteright that should be treated as an apostrophe:

```
lookup SUB_41 {
	@SUB_41_0_LA_0 = [A B C D E Eacute F G H I J K L M N O P Q R S T U V W X Y Z a b c d e eacute
             f g h i j k l m n o p q r s t u v w x y z];
	sub quoteright.alt' lookup SUB_40 @SUB_41_0_LA_0;
	@SUB_41_1_BT_0 = [colon colon.alt comma exclam exclam.alt period question question.alt semicolon 
             semicolon.alt];
	sub @SUB_41_1_BT_0 quoteright.alt' lookup SUB_40;
} SUB_41;
```

(where `lookup SUB_40` is simply a substitution of `quoteright.alt` for `quoteright`). I knew at the time that this was imperfect, but on the other hand I set an entire book in Junicode with `ss18` on, and it required very little manual intervention.

I haven't thought about this problem in years, but this seems a decent enough platform to build on.

**Comment by Alice196498** [comment]
It is a bit complex, but I think it’s quite poſſible to get ſomething that works in moſt caſes with OpenType. ( and ) don’t traditionally take any ſpace, but otherwiſe, this does look like a good platform to build on and improve upon.

**Comment by psb1558** [comment]
The most difficult problem (really the *only* difficult problem) is that of the em-space at sentence end. To do this, I have experimentally added a character space.ss18 which has the width of the em-square minus the right sidebearing of the period. ss18 does a contextual substitution:
```
  sub period space' @AllCapitals by space.ss18 ;
```
This works well in InDesign, Affinity Publisher, LibreOffice Writer, LuaLaTeX, Firefox, and Chrome. It *does not* work in Safari because some applications work by reading the width of the space out of the font instead of "printing" the space. (I may be able to test in MS Word and Edge later.)

An alternative is to create a character period.ss18 with extra space added to the right sidebearing and use that at sentence end. The problem comes when the end of a sentence coincides with the end of a line. Then you get a big ugly space messing up the right-justification, and as far as I know there is nothing to do about that.

The advantage of the alternate space, on the other hand, is that applications eat the space at line-end, so justification is never disrupted.

Is a solution that works in most apps but not all good enough?

**Comment by Alice196498** [comment]
For what reaſon would you not in ſuch a circumſtance ſubſtitute the word-ſpace with the em quad, rather than creating an entirely new ſpace glyph ? I’ve found that to work rather well with ſome RegEx that I uſe as a ſimilar ſolution, but it can, at times, cauſe iſſues. One trouble I ſee with that ſimple replacement is the caſe of abbreviations, like Dr. or Mrs., &c.

I have ſeen that ſome applications read the width of the ſpace, which is very ſtrange, and not a behaviour I like. Your ſolution with adding ſpace to the right of the character does have that iſſue, and it’s one reaſon I never did that when making edits to typefaces. I tried it after reading in a printing manual that it may be recommended to caſt certain characters with extra ſpace, and thought I’d try it, as it ſounded like a good idea, but that only works where a thin ſpace is wanting : namely, in the caſe of before the colon, ſemi-colon, and marks of interrogation and exclamation, and around quotation marks. Extra ſpace after muſt be handled differently, leſt you invite that ſame problem you encountered here, plus others.

Theſe are ſome reaſons that I take great iſſue with the modern trend of uſing word-ſpaces in all ſituations : it cauſes great difficulty for automated ſolutions to[ dealing with ſentences,](https://widespacer.blogspot.com/2012/12/one-space-is-just-wrong.html) ſcreen-reading, machine tranſlation, and [multitudinous](https://widespacer.blogspot.com/p/blog-page.html) other ſituations that are alleviated by ſimply ſpacing properly in the firſt place. I have managed ſolutions for this uſing Regular Expreſſions, but even that is imperfect ; and while ſuch a ſolution could be copied over uſing OpenType ſubſtitutions, and I am happy to provide what I have, it does need a little more fine-tuning in certain caſe, ſuch as abbreviations, for example, whereof I have not accounted for as many as I ſhould like.

All that being ſaid, all of theſe ſolutions for automatick replacements will only work in moſt apps—there are apps that are not OpenType aware, and cannot toggle the features at all—for theſe, “ always on ” variants can help, but even then, you’ll meet certain rare applications that ignore that as well. The beſt we can do is borrow clever yet weighty ſolutions ; then labour heavily that they may ſeem light. Working as ſuch, it is poſſible to allow the coverage of nearly all ſtandard caſes, and many, perhaps even a majority of, edge caſes, and it is my opinion that doing ſo is a worthy endeavour, deeply appreciated by all who find need or want to make uſe of the tools we can provide them.

**Comment by psb1558** [comment]
I'll be honest here: I simply forgot about the em quad—or better, the emspace (U+2003), which has the advantage that apps can expand or condense it as they would a space. But after a little experimentation, I lean towards using a variant of the space, though I take your point about the advantages of having different characters for different functions.*

One problem is that not all apps treat the emspace properly. It should behave like a space in that the app should eat it at line-end, but not all do. (So far I've only observed a problem in Affinity Publisher, but that's a warning that other problem apps may be out there.) If using the emspace solved the problem in Safari, that would be a big argument in its favor, but it does not.

In any case I would want a variant of emspace. My period.ss18 compensates for the right sidebearing of the period: it is ninety to ninety-one percent of the width of the emspace, depending on the width of that sidebearing. I would make an emspace.ss18 with the same width as space.ss18.

*Side note: I spend a good bit of energy worrying about accessibility and searchability, and in that respect the situation here is interesting. If you search for a space in Adobe Acrobat, the app will find both space and emspace. Search for the emspace and the app will find only emspaces. (I like that behavior.) Apple Preview, on the other hand, won't search for either spaces or emspace: that is part of the general wretchedness of searching in Preview. In Firefox, a search for spaces will also find emspaces, but a search for emspaces won't find anything. In Chrome, searching for either spaces or emspaces will find both. (This isn't an argument for or against the emspace, but just an observation.)

Here, btw, is a sample with ss08, ss16, and ss18 all turned on, and the Space axis set to its maximum value.

<img width="785" alt="image" src="https://user-images.githubusercontent.com/16215767/180896748-6ece375e-8bbf-4376-bce9-635fe4655a0a.png">

**Comment by Alice196498** [comment]
I agree with the em ſpace—I tend to call them the ſame thing, for that I initially learned theſe things through 18th century and 19th century works. It’s the character I uſe myſelf. I do find it curious that there are two variants of em and en ſpaces kept in Unicode, but there are reaſons for that. I am glad to hear your opinion in regard to the reaſoning for preferring a variant glyph here, when for the benefits provided, and to leſſen complexity, you choſe to aboliſh the variant longs.alt ; and that is ſomething I very much ſupport, as ſimplicity, where beſt ſuited, does tend to lead to a more elegant and workable deſign.

It is unfortunate that not all applications ſupport it properly, either due to the typefaces they chuſe lacking it, or poor handling of the characters themſelves. Both of theſe caſes are, in my opinion, the reſult of bad or wrong deſign, or, on occaſion, a very out-dated ſyſtem. It is my ſincere and true hope that more applications moving forward find better ſupport for theſe things, and keep the uſer in mind.

When accounting for things like this, it’s alſo important to conſider the ways the uſer may ſet their text—ſome good uſers will double ſpace, which is far more appropriate, and others will uſe an em ſpace or quad character : theſe ſituation ſhould alſo be conſidered by the ſtyle, ſo as not to be injurious to the work of the typeſetter, but yet allowing for the preference of the uſer as well.

Looking at your example, the ſpacing is rather even and good—I ſhould recommend more ſpace to be ſet after colons and ſemi-colons, however, as it appears to my eye to be rather cloſer to the word-ſpace than might be appropriate in thoſe lines. The ſpaces after ſentences, however, look quite nice, and do give ſufficient enough a break between them that the reader may retain and reflect upon what has juſt been read. This typeface, as it is coming along, is well ſuited to ſetting a great many works. If there may in future be the addition of the Greek and Cyrillick alphabets, it would then be ſufficiently able to ſet nearly any work required in Europe, the Americas, and parts of Africa ; furthermore, with the addition of Vietnameſe ſupport, parts of Aſia as well, both contemporarily and in an hiſtorical ſetting. That is a great achievement indeed.

Regarding your ſide note : I alſo ſpend time thinking about acceſſibility and ſearchability. Both of thoſe things are important to me, and important for people all around the world—eſpecially thoſe with diſabilities. It’s one reaſon why ſolutions like this can be ſo elegant and uſeful, and ſtandards, if created well and with the uſer in mind, are of uſch great utility and importance. The behaviour you’re ſeeing in theſe different applications is curious indeed ! I alſo like the behaviour in Adobe Acrobat, and think that particular method is beſt. I would that other applications were more inclined to follow its model.

**Comment by psb1558** [comment]
Grab fonts/variable/Elstob.ttf (not the italic) and give it a whirl. I know that a number of refinements are still needed.

<img width="728" alt="image" src="https://user-images.githubusercontent.com/16215767/180918359-da0606f5-3d3c-47a4-abef-bad8f28b152d.png">

**Comment by Alice196498** [comment]
I’ve been trying this out, and it ſeems to work rather well in general. It fails for ?’ !’ as well as ?” and !” and doeſn’t ſeem to add ſpace to them. One potential improvement that would fail leſs than it helps would be to let any ſingle capital letter followed by a period, ſpace, and capital letter (A. Brown, for example) be aſſumed an abbreviation, and not have a wider ſpace. This would cauſe a ſhorter ſpace in ſome rare caſes, but would help a vaſt many more. It alſo feels as if the added ſpace might be ſmaller than the thin ſpace—perhaps it’s ignoring the wide-ſpacing change, or perhaps it’s a trick of the ſetting and no problem at all. I do notice that after certain letters, ſuch as r or y, the period and comma appear to be kerned a little cloſe, which is eſpecially noticeable when quotation marks are involved, as they appear nearly directly up againſt the letters at times, as you can ſee in your image with “ party.” in the firſt line. It alſo ſeems that in caſes ſuch as “ readers’ ” there appears to be no way to cloſe up the apoſtrophe, and it is conſidered to be a quotation mark. It alſo is adding ſpace when a thin ſpace is already preſent, which may not be ideal. The French « and » may want to be added as quotation marks, and ¿ and ¡ for Spaniſh. If there were an interrobang, I ſhould recommend its incluſion here too, both as ‽, and as ⸘ ; however, given that the typefaces currently lacks theſe ſorts, it may be ſet aſide. I alſo notice that it may be uſing a word-ſpace after a colon and ſemi-colon not followed by a capital letter, when it ought to be uſing a wider ſpace, but not quite as wide as the ſpace wanted before a capital letter with a colon.

Theſe are the main things I’ve noticed ſo far, but as you ſaid, a number of refinements are ſtill needed.—Overall, it works wonderfully for a firſt iteration.

**Comment by psb1558** [comment]
I thought there wasn't supposed to be extra space after marks of punctuation. Should it be different for **?** and **!**, thus?

<img width="290" alt="image" src="https://user-images.githubusercontent.com/16215767/181085912-2355fb6a-e0e5-4fff-9e24-24a70f42c7d4.png">

The right sidebearings of **:** and **;** produced by ss18 are the difference between a space and an enspace—so you get a full enspace after those marks. I had adjusted the sidebearing for the colon, but not yet the semicolon. But now:

<img width="182" alt="image" src="https://user-images.githubusercontent.com/16215767/181086431-10cfa4cb-a67e-4cbd-a16b-64480d613f91.png">

My thin space may be too thin: I will adjust. I've been thinking about French « and » and other marks, and will add them after the rules are set.

As we're imitating metal type here, the modern kerning of quotation marks won't do (a qu. mark woudn't ordinarily overlap the character before). I'll have ss18 take care of this, with glyphs for which the kerning is looser or suppressed.

**Comment by Alice196498** [comment]
Yes, ſorry—? and ! have a thin ſpace ſet between them and the quotation marks. . , — don’t have ſpace after them, uſually, but ſometimes might be hair-ſpaced if required, or if the typeſetter prefers a hair-ſpaced em daſh.

Oh, all right. I was ſeeing that it looked a little ſmaller in juſtified text in the browſer, but juſtified text in the browſer doeſn’t uſe the beſt algorythm anyway ; ſo that could have been part of what I was ſeeing. Looking carefully at your image here, it appears that the ſpace before the ſemi-colon is ſlightly wider than the colon. I perſonally think this looks better, and ſhould recommend that the colon follow the ſemi-colon.

The ſpaces ſet here ſhould be equal to the fixed verſions, and the thin ſpaces ſhould all retain the ſame width, in line with the maſter ; thus, in the wide-ſpace verſion, all variants of thin ſpace ſhould retain 200 units of ſpacing, and em units 1000, with an en being 500, and the ſpace after a colon before a capitol would be 666—two thick ſpaces. Therefore, there ſhould be no difference between ſetting the ſpaces manually and toggling ss18, and ss18 ſhould alſo account for manual ſetting, thereby only affecting non-ſpaced text.

Imitating metal-type is a good way to look at it, I agree. I wouldn’t fully ſuppreſs the modern kerning, as it does often improve the readability and utility of the face, but looſening it in problematick places that are injurious to the text flow is a good idea.—I find that I’m often ſtopped by the period and clothing quote in faces that kern them ſo cloſely as if to make them a ſingle glyph. That being ſaid, they did have kerned ſorts in the days of metal-type, but thoſe were more breakable and harder to manage, as they’d either be caſt that way, or need to be ſhaved/cut to fit. I think we’re in agreement on this, baſed on what you’re ſaying here. As I underſtand it : This iſn’t meant to copy fuſil types ſo perfectly as to bring in their draw backs, but to imitate them as ſuch that their warm and good qualities may be repreſented accurately, while improving upon it where modern methods allow.

**Comment by psb1558** [comment]
I've fixed the spacing between exclam/question and following qu. marks: I just had to stop suppressing the substitution.

Colon and semicolon: good eye there. The colon and semicolon had the same left sidebearing, but the tail of the semicolon made it look at little wider. Now they are a visual (as opposed to arithmetical) match:

<img width="168" alt="image" src="https://user-images.githubusercontent.com/16215767/181105979-57281eaa-4db1-4abe-a30c-dda3c35361e1.png">

Yr third para: This is what I'm trying for. That said, the spacing e.g. between sentences, being always an em, looks best with the Spacing axis pushed all the way to 1; otherwise the various special spaces look too wide. For this reason, I don't know how well ss18 will do in the static fonts, which are what most people will use.

Kerning: one idea is to loosen the kerning between punctuation and right qu. marks on the wide end of the Spacing axis by about half:

<img width="727" alt="image" src="https://user-images.githubusercontent.com/16215767/181111362-dc6f2f54-8501-4a9d-85b2-7570ee96d41d.png">

Compare the same text with Spacing=0 and ss18 toggled off:

<img width="735" alt="image" src="https://user-images.githubusercontent.com/16215767/181111500-c8f55066-9442-4865-9627-78059827a332.png">

**Comment by Alice196498** [comment]
Oh, good.

For the colon and ſemi-colon : I thought it looked a bit off. If I recall correctly, in metal-type, they’d have to caſt them ſimilarly ſo as not to run into that ſame problem.

For the various ſpaces to look right in a low value for the ſpace axis, you’d need to account for the difference in ſpacing for all the ſpecial characters, I think. I ſhould think that a poſſible ſolution would be to note the ratios of ſpaces to each other at axis 1, and then give the ſame ratio to them, but caſt ſmaller for the axis at 0 ; this would give the ſame even appearance, while being more compreſſed to account for the modern ſenſibilities, while further ſtill allowing for the proper proportion in ſpacing.
 I’ve tried building ſome ſtatick founts for teſting, and the new long s implementation breaks in FontLab for ſome reaſon, and the build-ſcript may have ſome errors that prevent it from working correctly on Linux, as I was unable to uſe it. I ſhall make an iſſue for that. Were this not the caſe, I ſhould look into how it fairs in ſtatick verſions, and think about what could be done to help it. I did notice that it ſeems to make various inſtances : perhaps it would be good to have non-wide ſpaced inſtances and wide-ſpaced inſtances ? I was alſo conſidering ſuggeſting always-on variants for ſome of theſe ſtyles to fit with different ages of printing/needs, ſuch as one with ss08, ss18, ct in the liga-table rather than dlig, for example, to emulate 18th century printing, or ss08, ss16, and ss18 as in your example from earlier. How exactly ſomething like that might be done would be its own iſſue, I ſuppoſe, as it could potentially be ſomewhat involved.

Regarding the kerning : I do notice in your image that ? and ! appear ſomewhat cloſer to the matter than the apoſtrophes and turned commas, but otherwiſe, where quotation marks are meeting with punctuation, and that ſelf-ſame punctuation with its matter meet, this new verſion does look much better to my eye. It might be good to ever ſo ſlighly looſen it further in caſes of y, r, and related letters, as they ſtill feel ſlightly tighter than might be ideal, in my opinion, but this is much more a ſubtle thing.
 That being ſaid, as you ſay, this is one idea : I wonder what other ideas you have, becauſe this is a curious thing, and I amn’t ſure what all ſolutions are available.

I ſhould alſo like to make an addendum regarding the ſpace between a mark of exclamation and apoſtrophes, ! ” : while ſome printers would ſet a ſpace between them, either a thin ſpace or even a hair ſpace, others would ſet no ſpace between them. I generally ſet with a thin ſpace or a hair ſpace between the mark of exclamation and apoſtrophe, as that ſeems to be the moſt common and readable practice. That ſaid, both ways may be acceptable, depending on your ſenſibilities ; thus, we may ſet as follows :—
“ word ”
“ word ; ” or “ word ; ” “ word,” ‘ word ’,
“ word : ” or “ word : ” “ word.” ‘ word ’.
“ word ! ” or “ word ! ” “ word ” ; or “ word ” ; (colon follows ſemi-colon)
“ word ? ” or “ word ? ” “ word ” ?
The left is ſet with thin ſpaces, and the right with hair ſpaces. The right-moſt ſhews a comma and period in ſtandard as well as new Britiſh ſtyle, and alſo a when points fall outſide a quotation for ſome reaſon. Perhaps ſome table is a good idea, but you may alſo ſee ſome fun examples in [this book,](https://books.google.com/books?id=mbw9AAAAYAAJ&pg=PA63) which ſhews many examples of good ſpacing.

**Comment by psb1558** [comment]
Now ss18 is working with the Spacing axis. You can see the effect [here](https://github.com/psb1558/Elstob-font/blob/master/docs/Elstob-ss18-test.pdf). Briefly: when Spacing is 0, ss18 adds *some* extra space, and as Spacing approaches 1, the extra space approaches the full prescribed width. You only get the compleat 18th c. effect when Spacing=1, but along the way from 0 to 1 the punctuation acquires more of an old fashioned vibe when ss18 is on.

Kerning: for now, anyway, marks of punctuation with extra space are not kerned on the side with the space. Right single and double quotes are kerned differently with ss18 on, but only after period, comma, colon and semicolon.

I have no idea why the font is not working in FontLab. I can't test there. It works wherever I can test: InDesign, FontGoggles (and other apps with HarfBuzz layout), Chrome, etc.

**Comment by Alice196498** [comment]
Oh, wow—that’s a great demo : it juſt gets more readable as you move down. I do like the ſolution you’ve come up with here. It very well preſerves the proportion of ſpaces in regard to the word-ſpace, and the ſlight change to the leading is a nice touch.

Kerning : That’s fine—other adjuſtments can be made later ; it is important and helpful to note them, which is why I did.

Not working in FontLab : As far as I can tell, it ſeems to me that this is likely due to the way FontLab compiles OpenType features. It ſeems to compile it in a way that upſets the ſubſtitutions, cauſing many of the rules to break. I’m gueſſing that it’s either reordering things ſomewhere, or performing ſome kind of ſimplification that is being confuſed by the code, and making an error which diſallows them to compile as intended. The typeface itſelf works, but the OpenType compilation breaks ſome ſubſtitutions. At leaſt, that’s what happens in FontLab 7, it does alſo give ſome errors when building, ſuch as `WARNING: Class '@_uni16BA' contains no glyph and will be removed!` and `WARNING:fontTools.varLib:glyph .notdef has incompatible masters; skipping`, but it will nevertheleſs build the typeface. 8 breaks the ſubſtitutions and alſo gives fails on attempting to build, which is curious. Theſe are with the compiled typeface—trying to import the glyphs files directly cauſes other errors, and generally does not work. I believe that this is moſt likely a FontLab problem, although curious, and not ſomething wrong with your code.

**Comment by psb1558** [comment]
The italic plus the the latest version of the roman are available in fonts/variable. Also a [sample document](https://github.com/psb1558/Elstob-font/blob/master/docs/Elstob-Italic-ss18-test.pdf) for the italic. Italic spacing is always tricky, so this will have to be carefully reviewed.

BTW, I see that the kerning of the apostrophe is kind of insane. I'll work on fixing that.

**Comment by Alice196498** [comment]
This is looking very good ſo far. I ſee that you’ve added ſome common titles which are commonly ſet with a period ; I’ve made a liſt of ſome more, which, if included, would cover all, or nearly all, caſes in Engliſh, and French ſeems to be included here already—I ſhould alſo like to recommend ſetting a ſingle capital letter followed by a period with the word ſpace : while it is poſſible in that ſuch a caſe may be the end of a ſentence, this is very uncommon, and it is very common to ſee in names, as well as caſes ſuch as N. B. for _Nota Bene._
 Here is my liſt for the period :—
v., vs.
Mt.
Hon.
Rt.
Jr.
Mons, Mſgr.
Rev.
Gen.
Col.
Cmdr.
Maj.
Capt.
Lt.
MSgt.
TSgt.
SSgt.
Sgt.
Cpl.
Pvt.
Fr.
Pr.
Br.
Ven.
Emi.

I alſo have a couple for the apoſtrophe, which ſhould help in the rare caſes that it appears at the end of words, and fo’c’ſtle, which is odd that it’s being ſpaced in the middle :—
’n’
in’ (dialect form of ing)
fo’c’ſtle
ol’

The only other caſes that I can think of that wouldn’t work correctly are where s’ is involved. For the above, theſe are much leſs common to find as the ends of words, which would allow for them to not harm quoted matter, but s does end many words. Whether it’s beſt to include it as an exception and become always an apoſtrophe, or not and aſſume it’s a ſingle ending quote is a matter for conſideration. Both options have merits. Aſſuming it’s an apoſtrophe always would, in the uſual ſtyle, be, perhaps, leſs problematical, as it is more rare to meet with ſingle quotes in the firſt place ; however, given that we have the modern Britiſh ſtyle to contend with, ſuch a ſolution may potentially injurious to the common quoted matter for people who ſet their text as ſuch. Aſſuming it’s a cloſing quotation mark has the ſame problem in the oppoſite direction. A potential ſolution would be to allow this to be marked by the uſer by way of the ZWNJ.

I’ve alſo come acroſs an iſſue with the em ſpace : it ſeems that both the ſingle cloſing quote/apoſtrophe, as well as the double cloſing quote followed by a capital as a ſentence, thus treating ’ and ” as if they were periods. So, ſomething like jumpin’ Dogs would be ſet as jumpin’ Dogs, which is clearly not intended. The ſame occurs with ”.

It alſo ſeems that period followed by ) and ], ſpace, and a capital ; and period followed by a ſpace, ( or [, and a capital are both not included as ſentences, when they ought to be. (Thus, neither this ſentence nor the following, were one written, would be ſpaced appropriately.)

**Comment by psb1558** [comment]
I think I've got enough abbreviated titles (given that one can't catch all cases, where  do you draw the line?). And I'm not yet persuaded that a rule involving cap + period wouldn't cause nearly as many problems as it solves. And an apostrophe in sequences like **in’** (an English-specific thing) calls for the introduction of language tags for this feature, and I'd rather not go there if I can avoid it. We're handling the vast majority of cases now. The documentation will warn that the feature can't be fully automated and provide simple and reliable way to deal with the inevitable exceptions.

On the other hand, I will certainly deal with the bugs you identify in your last two paras. I've put up new builds of roman and italic in fonts/variable and a new italic sample doc with kerning fixes. (Kerning is a slow and iterative process: you fix one problem and instantly see two more.)

**Comment by Alice196498** [comment]
For abbreviated titles, I’d include theſe, as that’s the majority of common caſes you’ll meet acroſs languages, as far as I know. Outſide of that, I can think of very few that would be exceptions, which would lead me to believe that they’re either extremely uncommon, non-ſtandard, or more a more obſcure rule of ſome language. As for the cap + period + ſpace + cap iſſue, I uſe it in my own RegEx replacements, and have ſeen almoſt no caſes of that particular rule failing or cauſing a trouble—none immediately come to mind, at leaſt in general reading and looking through various texts. In moſt languages being dealt with here, it’s rather uncommon for a ſingle capital letter followed by a period, ſpace, and another capital letter to be a ſentence, rather than an abbreviation or title of ſome ſort. If you wanted to account for all-caps writing, let an exception ſtand for cap + cap + period + ſpace + cap + cap, which would then reduce the failure rate further. At that point, you’re left with either languages that have ſpecial needs, which are beſt ſet manually anyway, very rare caſes that need intervention, or caſes that ought not be ſet in a typeface like this in the firſt place, ſuch as code. That being ſaid, I’d like to hear of any counter-examples that would be injured by ſuch a rule. I uſually try to account for as much as I can with the ſimpleſt ruleſets with things like this ; ſo exceptions or broken lookups are always fun to find. While it’s true that not all caſes can be covered, and a line muſt be drawn ſomewhere, I believe that it’s beſt to cover what general caſes can be inaſmuch as poſſible, without introducing many hundreds of exception caſes. That belief is alſo the chief motivation for my recommending theſe further titles to be included : While it is more to add, thoſe are quite common, and the addition of them would further remove the need for ſpecial intervention on the part of the uſer greatly, and at the ſame time introduce no great iſſues with the overall ruleſet.

Regarding in’, I agree that it’s beſt to avoid a language tag ; thus, it’s a fair point that I concede.

I’m glad I could help find ſome new bugs, and alſo noticed a noticed two more : It ſeems that the colon and ſemi-colon, when followed by ſingle and double cloſing quotes are not ſpaced out appropriately from the matter, and the ſpace after them is not increaſed.

You’re right about kerning being a ſlow and iterative proceſs. It’s one of the more time-conſuming, tedious, and difficult things when making a typeface, and can be incredibly difficult to get right : doubly ſo in Italick.

**Comment by psb1558** [comment]
What is better?
<img width="638" alt="image" src="https://user-images.githubusercontent.com/16215767/181636157-f5084ee8-462e-4c0c-9357-c2264aeb3170.png">
or
<img width="556" alt="image" src="https://user-images.githubusercontent.com/16215767/181636247-b3b7280c-d88e-4790-ad1e-c4920ce73f7c.png">
If somewhere in between, I should be able to handle it with kerning. The other bugs are fixed.

**Comment by psb1558** [comment]
Quotes with guillemot, etc., are insane, since often the same Unicode character can go on the left or right. So there had to be a significant rewrite with contextual rules. But it seems to work:

<img width="591" alt="image" src="https://user-images.githubusercontent.com/16215767/181641490-fc72ae41-16e7-4840-b5aa-8b41458b8ebb.png">

**Comment by Alice196498** [comment]
For which is better, they both have merits, and both were done. In the top, I’d recommend having the apparent ſpace between the colon/ſemi-colon and quotation mark be equal to the apparent ſpace before it, which ſhould be equal in appearance to the thin ſpace glyph when the ſpacing axis is ſet to 1. As for which is beſt, I think a better queſtion is whether conſiſtency is moſt important, and then to look at whether a thin ſpace or hair ſpace is uſed in thoſe caſes. Perſonally, I tend to favour a thin ſpace, and would generally recommend it for the ſake of general conſiſtency and ſimplicity, but I have ſeen people who favour a hair ſpace as well. That would mean we make take the following as a rule : ſet a thin ſpace before ; : ! ? ” and ’ always, except when a period, comma, or daſh comes before ’ or ” ; likewiſe, ſet a thin ſpace after ‘ and “ always, except when a period, comma, or daſh is preſent there.

Regarding the quotes with guillemot, &c. : I agree, they can be a bit complex and difficult to ſet a fixed rule for becauſe of being the ſame character for each, and different languages having different preferences for how to ſet them. I admire and like what you’ve done here, and it looks like moſt of the iſſues are ſolved. I’ve been teſting the verſion I have, and other than the bugs that have been fixed, the main trouble I notice often is the ſingle capital thing, which is rather common, eſpecially with caſes like James A. Garfield, and the abbreviated titles that aren’t included. I ſhall continue to teſt and read with this, and note anything further that ſtands out to me.

**Comment by psb1558** [comment]
Try the current roman version in fonts/variable. This one adds cap + period does *not* mark the end of a sentence; and much fiddling with the  ss18 code.

**Comment by Alice196498** [comment]
I’ve been trying it out ſince the upload, and the new code is working well. I’ve not ſeen any errors ſo far from the rules that are there.

I did notice that the ſpace before the exclamation point is ſlightly thinner than the ſpace before the ſame character when ſet with ss18 off and a thin ſpace between them. This is alſo true with the queſtion mark, but to a much ſmaller degree. For the queſtion mark, it’s about a 6 mm difference, and for the exclamation point, it’s about 16 mm—at leaſt, according to the ruler in InDeſign. That may be too fine a difference to worry about, but I thought I’d mention it, as I did notice it.
 Here is a ſcreenſhot : the top of each is with ss18 off and a thin ſpace ſet, and the bottom is ss18 with ſpacing ſet to 1 ; at the bottom, you’ll ſee a merged verſion of them :—
![Untitled-1](https://user-images.githubusercontent.com/30258857/181704842-96809832-a1c9-4a4d-ae61-16f503ed1527.png)
It’s intereſting to ſee how they ſit in relation to each other. It ſeems that this difference in ſpacing is what gives the even appearance, and why I was feeling that ? and ! ſeemed ſo much cloſer ſet with ss18 than with thin ſpaces without ss18. You can ſee this very much with ! ”, and even moreſo with ; ” and : ” where the ſpacing has a very different and uneven appearance before vs. after the point.


As far as I can tell, the only other problem I’m running into is titles like Ven., Rep., Sen., Mt., and the v. for _verſus_ common in legal things (and vs. in common parlance), as well as the various military ones, which are more multitudinous in number. Otherwiſe, this ſeems to be juſt about complete. I’d likely have to do a ſignificant amount of reading to find problems if the common abbreviated titles were added. Anything elſe further outſide of this could be dealt with uſing a ZWNJ ſet beſide the point to be ignored, which is a function that may be beneficial to add.

Thank you ever ſo much for being willing to create theſe ſtyle ſets, and for letting me help teſt and conſult upon them and their implementation.

**Comment by psb1558** [comment]
You have your additional titles. I was worried about a performance hit when ss18 was on, but it doesn't seem to be happening. New build of the Roman font in fonts/variable.

<img width="738" alt="image" src="https://user-images.githubusercontent.com/16215767/181746066-7d2d16a8-ef5b-4c1f-af18-e387e29c832b.png">

**Comment by Alice196498** [comment]
Thank you for adding them. I don’t ſee any performance hit with it on either.

I do ſee that you alſo fixed the kerning/bearings of theſe ſorts—they’re now almoſt perfectly in-line with the thin ſpaced verſions ; however, I ſhould like to remark that the queſtion mark and quotations, ſingle and double, are both ſlightly farther to the right (about 2 mm) than they’d need to be to preſerve the harmonious appearance of the thin ſpaced. Otherwiſe, I think that’s about everything I’ve noticed that needs attention or correction, unleſs you have any further ſuggeſtions, ideas, or opinions. I am curious whether you have any thoughts regarding the ellipſis.

**Comment by psb1558** [comment]
Currently the thin space is simply added to the width of ? and ! (and also ¿ and ¡), but I will subtract the value of the adjacent sidebearings. After a final review, I will post a build of both roman and italic.

My only thought about the ellipsis is that the Unicode ellipsis is a barbarism. But you probably mean how to space it. I know of  some examples in late 18th c. printing. I'll look at them when I have a chance.

**Comment by Alice196498** [comment]
Looking at the example image, you can ſee that the ruler line touches the top of the quotation mark and the queſtion mark with the thin ſpace, but the bottom with ss18 : my recommendation is to copy this touching of the top, as that is about where the viſual centre is, and what allows them to ſit with even appearance. I alſo recommend that ¿ and ¡ match this proportion in the oppoſite direction, as they, too, are off in the ſame way.

I agree that the Unicode ellipſis is awful. And yea, I was talking about the ſpacing. It’s quite intereſting to note the hiſtory of ellipſis in print—there are very few examples in the 18th century of dots being uſed, as they uſually uſed a daſh, or ſometimes aſtreriſks or multiple daſhes. There are a couple of examples in Engliſh and French that I’ve ſeen, but they’re very rare and vary irregular in ſtyle when it comes to the uſe of dots for the ellipſis. In the 19th century, they ſtarted uſing dots a bit, but the uſe was ſtill a bit non-ſtandard till the latter half of the century and early part of the next. At this time, it was moſt common to ſet with an em ſpace between them, but ſome people ſet with a word ſpace, and ſome, very rarely, a thin ſpace. I perſonally follow the habit of uſing an en ſpace myſelf, but conſider a thin ſpace to be permiſſable. They ought never be ſet to ſolidly bear againſt each other, which comes back to my diſlike of the Unicode ellipſis. That being ſaid, if you do come acroſs any examples in the 18th century, I’d like to ſee them, becauſe they’re a fun treat in the hiſtory of uſe and ſenſibility of printers.

Speaking of daſhes : I ſhould perhaps recommend the kerning of em daſhes that ſtand beſide each other to be ſolid in order to allow the ſetting of even rules. ———, for example, would be a 3-em-rule, and ſhould appear as a ſingle line.

**Comment by psb1558** [comment]
I was misled by ellipses in a modern edition Boswell's Life of Johnson. They weren't in the first printing, but only used in the modern edition to signal abridgements. I don't think I'm going to do anything about the ellipsis. People can use Elstob's various spacing characters to produce whatever look they want.

**Comment by Alice196498** [comment]
Ooh, all right. I think that’s a good idea here, becauſe there are various ſenſibilities when it comes to the ſpacing of ellipſis.

**Comment by psb1558** [comment]
One could go on adding language and rules indefinitely, but for now I think this is done.

**Comment by Alice196498** [comment]
I’ve noticed that ss18 is adding ſpace around a colon when uſed in date/time formats. It might be a good idea to fix that, but otherwiſe, I’ve been uſing it, and it does ſeem moſtly done aſide from thoſe uncaught edge caſes. It might alſo be a good idea to add a bit of additional ſpace to ’ and ” after an f, as currently there is little apparent difference between ss18 being on or off in the caſe of that paring.

**Comment by psb1558** [comment]
These things should be easy to fix.
